### ISSUE TITLE
[WIP] NN Optimization Ideas

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#52]
I'm going to use this issue as a scratch-pad to note down ideas which could improve performance of our models. I'll update this post as I come up with more thoughts.

1. The weight_scale (the standard-deviation of the Gaussian used to initialize the weight matrices) is not a hyperparameter. It should be. Networks with wrong weight_scale can't overfit small training sets (checking for such overfit is a good sanity check).
2. SingletaskDeepNetwork/MultitaskDeepNetwork should support arbitrary numbers of hidden layers. I'm not sure that these extra layers will improve performance, but I think we need to do the checks carefully.
3. SingletaskDeepNetwork/MultitaskDeepNetwork should support dropout/batch-normalization. Unfortunately, this is slightly complicated right now since batch-normalization serialization in keras is currently broken (https://github.com/fchollet/keras/pull/1518)
4. Currently, only SGD with various knobs is enabled. Adam, Adadelta, Adagrad, RMSProp should be enabled as well.

**Comment by rbharath** [comment]
fchollet/keras#1518 has been fixed, so batch-normalization layers should be added to ```SingletaskDeepNetwork```.

**Comment by rbharath** [comment]
Another couple ideas:

4. Currently, only SGD with various knobs is enabled. Adam, Adadelta, Adagrad, RMSProp should be enabled as well.

**Comment by rbharath** [comment]
Most of these fixes have been added via Tensorflow. Going to close for now; specific needs can be addressed in follow-on PRs.
